### ISSUE TITLE
[clang-tidy] False positive for modernize-avoid-c-arrays in smart pointers and their make functions

### ISSUE THREAD

**Issue by kuzi117** [llvm/llvm-project#132260]
Clangd version: 19.1.2

Using the `T[]` specialization for `make_unique`/`make_unique_for_overwrite` and the resulting unique_ptr, clang-tidy reports that the template parameter should be replaced with `std::array`.

Example:

```c++
  std::unique_ptr<T[]> W = std::make_unique<T[]>(n);
  std::unique_ptr<T[]> X = std::make_unique_for_overwrite<T[]>(n);
  std::shared_ptr<T[]> Y = std::make_shared<T[]>(n);
  std::shared_ptr<T[]> Z = std::make_shared_for_overwrite<T[]>(n);
```

Every instance of `T[]` in the example is flagged as violating cppcoreguidelines-avoid-c-arrays/modernize-avoid-c-arrays and suggests using `std::array` instead, which doesn't seem feasible in the context.

**Comment by EugeneZelenko** [comment]
Could you please try 20 or `main` branch? https://godbolt.org should be helpful.

**Comment by nolange** [comment]
Ping? this warning is pretty much everywhere in my codebase

**Comment by NagyDonat** [comment]
@EugeneZelenko This issue can be reproduced on recent clang (trunk, 22.1, 21.1, 20.1): [godbolt link](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,selection:(endColumn:19,endLineNumber:3,positionColumn:19,positionLineNumber:3,selectionStartColumn:19,selectionStartLineNumber:3,startColumn:19,startLineNumber:3),source:'%23include+%3Cmemory%3E%0A%0Atemplate+%3Cclass+T%3E%0Avoid+f(int+n)%7B%0A++std::unique_ptr%3CT%5B%5D%3E+W+%3D+std::make_unique%3CT%5B%5D%3E(n)%3B%0A%7D'),l:'5',n:'0',o:'C%2B%2B+source+%231',t:'0')),k:44.94424939528422,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((g:!((h:compiler,i:(compiler:clang_trunk,filters:(b:'0',binary:'1',binaryObject:'1',commentOnly:'0',debugCalls:'1',demangle:'1',directives:'0',execute:'1',intel:'1',libraryCode:'0',trim:'1',verboseDemangling:'0'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:c%2B%2B,libs:!((name:boost,ver:'178'),(name:outcome,ver:trunk)),options:'',overrides:!(),selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1),l:'5',n:'0',o:'+x86-64+clang+(trunk)+(Editor+%231)',t:'0')),header:(),k:55.05575060471577,l:'4',m:33.333333333333336,n:'0',o:'',s:0,t:'0'),(g:!((h:output,i:(compilerName:'x86-64+clang+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:1,wrap:'1'),l:'5',n:'0',o:'Output+of+x86-64+clang+(trunk)+(Compiler+%231)',t:'0')),header:(),l:'4',m:15.942028985507244,n:'0',o:'',s:0,t:'0'),(g:!((h:tool,i:(args:'-checks%3Dmodernize-avoid-c-arrays',argsPanelShown:'0',compilerName:'x86-64+clang+(trunk)',editorid:1,fontScale:14,fontUsePx:'0',j:1,monacoEditorHasBeenAutoOpened:'1',monacoEditorOpen:'1',monacoStdin:'1',stdin:'',stdinPanelShown:'1',toolId:clangtidytrunk,wrap:'1'),l:'5',n:'0',o:'clang-tidy+(trunk)+x86-64+clang+(trunk)+(Editor+%231,+Compiler+%231)',t:'0')),l:'4',m:50.724637681159415,n:'0',o:'',s:0,t:'0')),k:55.05575060471577,l:'3',n:'0',o:'',t:'0')),l:'2',n:'0',o:'',t:'0')),version:4)

**Comment by vbvictor** [comment]
Closing in in favour of older issue
